pkgs/node/src/forkchoice.zig

```diff
@@ -904,3 +924,327 @@ test "forkchoice block tree" {
         try std.testing.expect(searched_idx == i);
     }
 }
+
+// Helper function to create a deterministic test root filled with a specific byte
+fn createTestRoot(fill_byte: u8) types.Root {
+    var root: types.Root = undefined;
+    @memset(&root, fill_byte);
+    return root;
+}
+
+// Helper function to create a ProtoBlock for testing
+fn createTestProtoBlock(slot: types.Slot, block_root_byte: u8, parent_root_byte: u8) ProtoBlock {
+    return ProtoBlock{
+        .slot = slot,
+        .blockRoot = createTestRoot(block_root_byte),
+        .parentRoot = createTestRoot(parent_root_byte),
+        .stateRoot = createTestRoot(0x00),
+        .timeliness = true,
+    };
+}
+
+test "getCanonicalAncestorAtDepth and getCanonicalityAnalysis" {
+    // ============================================================================
+    // COMPREHENSIVE TEST TREE
+    // ============================================================================
+    //
+    // This test creates a single tree that exercises ALL key scenarios:
+    //   1. FORKS      - Multiple children from one parent (C has children D and G)
+    //   2. MISSED SLOTS - Gaps in slot numbers (slots 2, 4, 7 have no blocks)
+    //   3. ORPHANS    - Non-canonical blocks that get pruned (G, H, I when finalized past C)
+    //
+    // Tree Structure:
+    //
+    //   Slot:  0      1      3      5      6      8
+    //         [A] -> [B] -> [C] -> [D] -> [E] -> [F]    <- Canonical chain (head)
+    //                        \
+    //                         [G] -> [H] -> [I]         <- Fork branch (becomes orphans)
+    //                        (s4)   (s6)   (s7)
+    //
+    //   Missed slots: 2, 4 (on canonical), 7
+    //
+    // Block Details:
+    //   A = 0xAA (slot 0, genesis)
+    //   B = 0xBB (slot 1, parent A)
+    //   C = 0xCC (slot 3, parent B)     <- FORK POINT, missed slot 2
+    //   D = 0xDD (slot 5, parent C)     <- missed slot 4 on canonical
+    //   E = 0xEE (slot 6, parent D)
+    //   F = 0xFF (slot 8, parent E)     <- HEAD, missed slot 7
+    //   G = 0x11 (slot 4, parent C)     <- FORK starts here
+    //   H = 0x22 (slot 6, parent G)
+    //   I = 0x33 (slot 7, parent H)
+    //
+    // Node indices in protoArray:
+    //   A=0, B=1, C=2, D=3, E=4, F=5, G=6, H=7, I=8
+    //
+    // ============================================================================
+
+    var arena_allocator = std.heap.ArenaAllocator.init(std.testing.allocator);
```

Comment:
just use `std.testing.allocator` in tests, because the arena allocator is good for performance, but it won't tell you if one particular allocation isn't freed.